### PR TITLE
firstboot: validate keymap entry

### LINKED ISSUE #30940: No validation checks performed on keymap entry in systemd-firstboot
### systemd version the issue has been seen with

254

### Used distribution

Gentoo

### Linux kernel version used

_No response_

### CPU architectures issue was seen on

x86_64

### Component

other

### Expected behaviour you didn't see

I expected that incorrect entries using characters rather than the number selection, i.e "uk" or "de" would be rejected on an incorrect entry, i.e "kjsdhf".

### Unexpected behaviour you saw

Instead of prompting the user to re-enter their keymap entry, it was instead written directly to KEYMAP, resulting in `KEYMAP=lksdjf`.

### Steps to reproduce the problem

1. Launch systemd-firstboot on a system that doesn't have a proper setup completed (alternatively, run `systemd-firstboot --reset` then `systemd-firstboot`).
2. Proceed normally to the keymap section
3. Use random characters that wouldn't be an accepted keymap.

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by edaigle**
As described in #30940, systemd-firstboot currently does not perform any validation on keymap entry, allowing nonexistent keymaps to be written to /etc/vconsole.conf. This commit adds validation checks based on those already performed on locale entry, preventing invalid keymaps from being set.

Closes #30940

### CODE REVIEW

**Comment by edaigle** [comment]
This is based off the validation check for the locale, which takes the root directory into account when deciding how strict to be in validating. I am new to the codebase and still not fully sure of the reason for this, if it isn't necessary in the case of the keymap I am happy to simplify this. Feedback very welcome!

File: src/firstboot/firstboot.c [hunk 1]
```diff
@@ -458,8 +458,20 @@ static int process_locale(int rfd) {
...
+static bool keymap_is_ok(int rfd, const char *name) {
+        assert(rfd >= 0);
+
+        return dir_fd_is_root(rfd) ? keymap_exists_bool(name) : keymap_is_valid(name);
+}
+
 static int prompt_keymap(int rfd) {
         _cleanup_strv_free_ char **kmaps = NULL;
+        bool (*is_valid)(const char *name) = dir_fd_is_root(rfd) ? keymap_exists_bool
+                : keymap_is_valid;
```
Comment (yuwata) [inline src/firstboot/firstboot.c]: Not a big matter, but it is not necessary to insert line break here. We allow 109ch per line.

**Review by yuwata — APPROVED** [review]

File: src/firstboot/firstboot.c [hunk 2]
```diff
@@ -458,8 +458,20 @@ static int process_locale(int rfd) {
...
+
+static bool keymap_is_ok(int rfd, const char *name) {
+        assert(rfd >= 0);
+
+        return dir_fd_is_root(rfd) ? keymap_exists_bool(name) : keymap_is_valid(name);
+}
+
 static int prompt_keymap(int rfd) {
         _cleanup_strv_free_ char **kmaps = NULL;
+        bool (*is_valid)(const char *name) = dir_fd_is_root(rfd) ? keymap_exists_bool
```
Comment (poettering) [inline src/firstboot/firstboot.c]: dir_fd_is_root() possibly an error here. which you treat as "true" here. Which you probably shouldnt.

I mean, it's fine to treat failures as true or as false sometimes, dependening on the case, but this should always be made explicit via `!= 0` or `> 0` checks, to underline you are aware this function doesn't just return a boolean.

Comment (poettering) [reply src/firstboot/firstboot.c]: (that said, we nowadays prefer if there's at least debug logging about unexpected errors. i.e. a quick log_debug_errno() line if the check returns < 0. similar for the keymap_exists() check above)

Comment (poettering) [inline src/firstboot/firstboot.c]: this ternary op you basically have in place twice. You could move the ternary op that determines the function into a function itself... i.e.

```c
static typeof(keymap_is_valid) determine_keymap_validity_func(int rfd) {
       int r;

       r = dir_fd_is_root(rfd);
       if (r < 0) 
               log_debug_errno(r, "Unable to determine if operating on host root directory, assuming we are: %m");

       return r != 0 ? keymap_exists_bool : keymap_is_valid;
}
```

File: src/firstboot/firstboot.c [hunk 3]
```diff
@@ -1695,6 +1707,8 @@ static int run(int argc, char *argv[]) {
         /* We check these conditions here instead of in parse_argv() so that we can take the root directory
          * into account. */
 
+        if (arg_keymap && !keymap_is_ok(rfd, arg_keymap))
```
Comment (poettering) [inline src/firstboot/firstboot.c]: with the above you'd call it like this:

```
if (arg_keymap && !determine_keymap_validity_func(rfd)(arg_keymap))
        …
```
you get the idea.

(untested, but I think this should just work)

File: src/firstboot/firstboot.c (hunk 2 above)
Comment (edaigle) [reply src/firstboot/firstboot.c]: To get it to build I had to do `static typeof(&keymap_is_valid)`, which I think makes sense as we are returning a function pointer - alternatively the normal ugly C fptr syntax 

`static bool (*determine_keymap_validity_func(int rfd))(const char*)`
works fine

**Comment by edaigle** [comment]
Thanks for the help, good points about handling the error explicitly and reducing duplication. Updated and (with the change to the signature) tested working

**Comment by poettering** [comment]
lgtm. thanks!

**Pull request merged**